File: examples/python/cart/handlers/remove_item.py
"""RemoveItem command handler."""

from datetime import datetime, timezone

from google.protobuf.any_pb2 import Any
from google.protobuf.timestamp_pb2 import Timestamp

from angzarr import angzarr_pb2 as angzarr
from proto import domains_pb2 as domains
from state import CartState

from .errors import CommandRejectedError


def handle_remove_item(command_book, command_any, state: CartState, seq: int, log) -> angzarr.EventBook:
    if not state.exists():
        raise CommandRejectedError("Cart does not exist")
    if not state.is_active():
        raise CommandRejectedError("Cart is already checked out")

    cmd = domains.RemoveItem()
    command_any.Unpack(cmd)

    if cmd.product_id not in state.items:
        raise CommandRejectedError("Item not in cart")

    item = state.items[cmd.product_id]
    item_subtotal = item.quantity * item.unit_price_cents
    new_subtotal = state.subtotal_cents - item_subtotal

    log.info("removing_item", product_id=cmd.product_id)

    event = domains.ItemRemoved(
        product_id=cmd.product_id,
        quantity=item.quantity,
        new_subtotal=new_subtotal,
        removed_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
    )

    event_any = Any()
    event_any.Pack(event, type_url_prefix="type.examples/")

    return angzarr.EventBook(
        cover=command_book.cover,
        pages=[angzarr.EventPage(num=seq, event=event_any, created_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())))],
    )

File: examples/rust/cart/src/handlers/remove_item.rs
//! Remove item command handler.

use angzarr::proto::{CommandBook, EventBook};
use common::proto::{CartItem, CartState, ItemRemoved, RemoveItem};
use common::{
    decode_command, now, require_exists, require_status_not, BusinessError, Result,
};

use crate::errmsg;
use crate::state::{build_event_response, calculate_subtotal};

/// Handle the RemoveItem command.
///
/// Removes an item from the cart entirely.
pub fn handle_remove_item(
    command_book: &CommandBook,
    command_data: &[u8],
    state: &CartState,
    next_seq: u32,
) -> Result<EventBook> {
    require_exists(&state.customer_id, errmsg::CART_NOT_FOUND)?;
    require_status_not(&state.status, "checked_out", errmsg::CART_CHECKED_OUT)?;

    let cmd: RemoveItem = decode_command(command_data)?;

    let item = state
        .items
        .iter()
        .find(|i| i.product_id == cmd.product_id)
        .ok_or_else(|| BusinessError::Rejected(errmsg::ITEM_NOT_IN_CART.to_string()))?;

    let removed_quantity = item.quantity;

    // Calculate new subtotal
    let items: Vec<CartItem> = state
        .items
        .iter()
        .filter(|i| i.product_id != cmd.product_id)
        .cloned()
        .collect();
    let new_subtotal = calculate_subtotal(&items);

    let event = ItemRemoved {
        product_id: cmd.product_id,
        quantity: removed_quantity,
        new_subtotal,
        removed_at: Some(now()),
    };

    Ok(build_event_response(
        state,
        command_book.cover.clone(),
        next_seq,
        "type.examples/examples.ItemRemoved",
        event,
    ))
}
